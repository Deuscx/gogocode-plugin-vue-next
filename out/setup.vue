
<script setup lang="ts">
import { reactive,ref,defineProps,defineEmits } from 'vue'
import Hello from 'hello.vue'
const props = defineProps({
  width: {
    type: Number,
    default: 446,
  },
  title: {
    type: String,
    default: '',
  },
})
const emit = defineEmit(['accepted'])
const a = ref(1)
const b = reactive([])

function fn() {
  return 1
}

function accept() {
  emit('accepted')
}
</script>
