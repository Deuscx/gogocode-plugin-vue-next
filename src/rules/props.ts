import type { GoGoAST } from 'gogocode'

export default function (ast: GoGoAST) {
  const script = ast.find('<script></script>')
  const opt = script.find('{props: $_$1}')

  opt.each((item) => {
    const value = item.match[1][0].value
    const node = item.match[1][0].node
    if (node.type === 'ObjectExpression')
      script.after(`const props = defineProps(${value})`)
  })

  script.find('props: $_$1').each((item) => {
    if (item.parent(2).has('defineComponent'))
      item.remove()
  })

  return script.root()
}
