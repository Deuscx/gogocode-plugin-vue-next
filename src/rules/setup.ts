import type $ from 'gogocode'
import type { Node } from '@babel/types'

export function insertDefineProps(script: $.GoGoAST) {
  const opt = script.find('{props: $_$1}')
  const hasProps = script.has('defineProps')
  let needToInjectProps = false

  opt.each((item) => {
    const value = item.match[1][0].value
    const node = item.match[1][0].node
    if (node.type === 'ObjectExpression') {
      needToInjectProps = true
      script.after(`const props = defineProps(${value})`)
    }
  })

  script.find('props: $_$1').each((item) => {
    if (item.parent(2).has('defineComponent'))
      item.remove()
  })

  if (!hasProps && needToInjectProps) {
    script.replace(
      'import { $$$1 } from \'vue\'',
      'import { $$$1, defineProps } from \'vue\'',
    )
  }
}

export function insertDefineEmits(script: $.GoGoAST) {
  const setup = script.find('{ setup() {$$$0} }')
  const emits = script.find('{ emits: $_$1 }')
  const hasEmits = script.has('defineEmits')
  let needToInjectEmits = false
  script.find('setup($_$,$_$1) {}').each((item) => {
    const match = item.match[1][0]
    const node = match.node
    const nodeType = node.type
    if (nodeType === 'Identifier') {
      // ep: setup(_, context)
      const contextName = match.value
      setup.each((item) => {
        item.replace(`${contextName}.emit`, 'emit')
      })
    }
    else if (nodeType === 'ObjectPattern') {
      // ep: setup({}, {emit})
    }
  })

  // Remove emits
  const emitsArr = emits.match?.[1]?.[0].value
  if (emitsArr) {
    script.after(`const emit = defineEmit(${emitsArr})`)
    script.find('emits: $_$1').each((item) => {
      if (item.parent(2).has('defineComponent'))
        item.remove()
    })

    needToInjectEmits = true
  }

  if (!hasEmits && needToInjectEmits) {
    script.replace(
      'import { $$$1 } from \'vue\'',
      'import { $$$1, defineEmits } from \'vue\'',
    )
  }
}

export function exposeSetup(script: $.GoGoAST) {
  script.find('{ setup() {$$$0} }').each((item) => {
    // remove return statement
    (item.match.$$$0 as unknown as Node[]).forEach((node) => {
      if (node.type !== 'ReturnStatement')
        script.after(node)
    })
  })
  script.find('setup(){}').remove()

  // remove component
  const components = script.find('{ components: $_$1 }')
  const componentsArr = components.match[1][0].value
  if (componentsArr) {
    script.find('components: $_$1').each((item) => {
      if (item.parent(2).has('defineComponent'))
        item.remove()
    })
  }

  // remove defineComponent
  script.find('export default $_$1').each((item) => {
    item.remove()
  })

  // remove import defineComponent
  script.replace(
    'import { $$$1,defineComponent } from \'vue\'',
    'import { $$$1 } from \'vue\'',
  )
}
export function transformSetup(ast: $.GoGoAST) {
  const isSetupScript = ast.has('<script setup></script>')
  if (isSetupScript)
    return ast.generate()

  const script = ast.find('<script></script>')

  // 1. defineProps
  insertDefineProps(script)
  // 2. defineEmits
  insertDefineEmits(script)
  exposeSetup(script)
  return ast.generate().replace('script', 'script setup')
}
