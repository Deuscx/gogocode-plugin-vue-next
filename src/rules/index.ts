import type { GoGoAST } from 'gogocode'
import props from './props'

interface IRule {
  name: string
  rule: (ast: GoGoAST, api?: any, options?: any) => any
  test: RegExp
}

const rules: IRule[] = [{ name: 'arrayRefs', rule: props, test: /\.vue$/ }]
export default rules
