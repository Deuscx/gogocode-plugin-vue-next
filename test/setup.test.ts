import { resolve } from 'path'
import { describe, expect, it } from 'vitest'
import $ from 'gogocode'
import { transformSetup } from '../src/rules/setup'

describe('should', () => {
  it('setup', () => {
    const ast = $.loadFile(resolve(__dirname, './examples/setup.vue'), {
      parseOptions: { language: 'vue' },
    })

    // finally: add script setup
    expect(transformSetup(ast)).toMatchSnapshot()
  })

  it('props', () => {
    const ast = $.loadFile(resolve(__dirname, './examples/props.vue'), {
      parseOptions: { language: 'vue' },
    })

    // finally: add script setup
    expect(transformSetup(ast)).toMatchSnapshot()
  })

  it('emits', () => {
    const ast = $.loadFile(resolve(__dirname, './examples/emits.vue'), {
      parseOptions: { language: 'vue' },
    })

    // finally: add script setup
    expect(transformSetup(ast)).toMatchSnapshot()
  })
})
