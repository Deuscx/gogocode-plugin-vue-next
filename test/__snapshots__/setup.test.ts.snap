// Vitest Snapshot v1

exports[`should > remove setup return 1`] = `
"
<script lang=\\"ts\\">
import { defineComponent, reactive, ref } from 'vue'

export default defineComponent({
  name: 'Test'
})
const a = ref(1)
const b = reactive([])

function fn() {
  return 1
}
</script>
"
`;
